compiler/src/iree/compiler/Dialect/Flow/TransformExtensions/FlowExtensions.cpp

```diff
@@ -139,8 +140,72 @@ static void rewriteExtractSlices(PatternRewriter &rewriter,
   });
 }
 
+static void cloneOpsIntoForeachThreadOp(RewriterBase &rewriter,
+                                        scf::ForeachThreadOp foreachThreadOp) {
+  // 1. Find all ops that should be cloned into the ForeachThreadOp.
+  llvm::SetVector<Value> valuesDefinedAbove;
+  mlir::getUsedValuesDefinedAbove(foreachThreadOp.getRegion(),
+                                  valuesDefinedAbove);
+  // Add all ops who's results are used inside the ForeachThreadOp to the
+  // worklist.
+  llvm::SetVector<Operation *> worklist;
+  for (Value v : valuesDefinedAbove)
+    if (Operation *op = v.getDefiningOp()) worklist.insert(op);
+  llvm::SmallVector<Operation *> opsToClone;
+  llvm::DenseSet<Operation *> visited;
+
+  // Process all ops in the worklist.
+  while (!worklist.empty()) {
+    Operation *op = worklist.pop_back_val();
+    if (visited.contains(op)) continue;
+    visited.insert(op);
+
+    // Do not clone ops that are not clonable.
+    if (!mlir::iree_compiler::IREE::Flow::isClonableIntoDispatchOp(op))
+      continue;
+
+    // Do not clone ParallelInsertSliceOp destinations.
+    bool isDestination =
+        any_of(foreachThreadOp.getTerminator().getYieldingOps(),
+               [&](Operation &insertOp) {
+                 return cast<tensor::ParallelInsertSliceOp>(&insertOp)
+                            .getDest()
+                            .getDefiningOp() == op;
+               });
+    if (isDestination) continue;
+
+    opsToClone.push_back(op);
+
+    // Add all operands to the worklist.
+    for (Value operand : op->getOperands()) {
+      Operation *operandOp = operand.getDefiningOp();
+      if (!operandOp) continue;
+      worklist.insert(operandOp);
+    }
+  }
+
+  // 2. Clone ops and replace their uses inside the ForeachThreadOp.
+  OpBuilder::InsertionGuard g(rewriter);
+  rewriter.setInsertionPointToStart(
+      &foreachThreadOp.getRegion().getBlocks().front());
+  for (Operation *op : llvm::reverse(opsToClone)) {
```

Comment:
This might run into issues. Previously doing this this wasnt enough because of general DAG dependency. It was more stable to do a topological sort of the ops that are to be cloned and then inlining then in that order....

Reply:
Ok, if this is actually not good enough, I'll add the topological sort in a follow up PR. I'm just testing with a small example at the moment (`tests/e2e/models/fullyconnected.mlir`) and will expand to larger ones once I get that one running.